OTAnalytics/plugin_filter/dataframe_filter.py

```diff
@@ -0,0 +1,162 @@
+from __future__ import annotations
+
+from datetime import datetime
+from typing import Iterable, Optional
+
+from pandas import DataFrame, Series
+
+from OTAnalytics.domain.filter import (
+    Conjunction,
+    Filter,
+    FilterBuilder,
+    FilterBuildError,
+    Predicate,
+)
+
+
+class DataFrameConjunction(Conjunction[DataFrame, Series]):
+    """Represents the conjunction of two DataFrame predicates.
+
+    Args:
+        first_predicate (Predicate[DataFrame, Series]): first predicate to
+            conjunct with
+        second_predicate (Predicate[DataFrame, Series]): second predicate to
+            conjunct with
+    """
+
+    def __init__(
+        self,
+        first_predicate: Predicate[DataFrame, Series],
+        second_predicate: Predicate[DataFrame, Series],
+    ) -> None:
+        super().__init__(first_predicate, second_predicate)
+
+    def test(self, to_test: DataFrame) -> Series:
+        return (self._first_predicate.test(to_test)) & (
+            self._second_predicate.test(to_test)
+        )
+
+    def conjunct_with(
+        self, other: Predicate[DataFrame, Series]
+    ) -> Predicate[DataFrame, Series]:
+        return DataFrameConjunction(self, other)
+
+
+class DataFramePredicate(Predicate[DataFrame, Series]):
+    def conjunct_with(
+        self, other: Predicate[DataFrame, Series]
+    ) -> Predicate[DataFrame, Series]:
+        return DataFrameConjunction(self, other)
+
+
+class DataFrameFilter(Filter[DataFrame, Series]):
+    def __init__(self, predicate: Predicate[DataFrame, Series]) -> None:
+        """A `DataFrame` filter.
+
+        Args:
+            predicate (Predicate[DataFrame, Series]): the predicate to test the
+                DataFrame against.
+        """
+        super().__init__(predicate)
+
+    def apply(self, data: Iterable[DataFrame]) -> Iterable[DataFrame]:
+        return [datum[self._predicate.test(datum)] for datum in data]
+
+
+class DataFrameIsWithinDate(DataFramePredicate):
+    """Checks if the DataFrame rows are within a date range.
+
+    Args:
+        column_name (str): the DataFrame column name to apply the predicate to
+        start_date (datetime): the start date of the date range
+        end_date (datetime): the end date of the date range
```

Comment:
See comment below about inclusive and exclusive dates. The documentation should state the final decission.